rfc/plugins.md

```diff
@@ -0,0 +1,235 @@
+compose and combine multiple plugins to fine tune their deployment
+process.
+
+### Identifying plugins
+
+Plugins are published to npm as ember-cli addons. They will also have
+the keyword ``ember-cli-deploy-plugin``.
+
+Any plugin that is installed will be loaded by default, similar to
+ember-cli addons.
+
+### Base plugin object
+
+Plugins will extend a base object provided by ember-cli-deploy. The API
+of this object will be small to start, but it will give us an easy place
+to give these plugins additional interfaces if needed.
+
+```javascript
+S3Plugin = DeployPlugin.extend({
+  ...
+});
+```
+
+### Hooks
+
+Plugins hook into the deployment pipeline with function names that match
+the steps in the pipeline process. These hooks will scoped under a
+special hooks object inside the plugin so that any private or plugin
+functions do not clash with any hook names that we add in
+the future.
+
+Although the hook functions are inside the hook object, they will be
+executed with a `this` binding that points to the plugin object.
+
+```javascript
+S3Plugin = {
+  hooks: {
+    assets: function() {
+      this.something();
+      this.uploadToS3();
+    })
+  },
+
+  something: function() { ... },
+
+  uploadToS3: function() { ... }
+}
+```
+
+### Provided properties
+
+Ember-cli-deploy will inject a number of properties onto each plugin.
+
+```javascript
+S3Plugin = {
+  ui: { ... },
+  appConfig: { ... },
+  deployConfig: { ... },
+  pluginConfig: { ... }
+}
+```
+
+Property | file | info
+--- | --- | ---
+`ui` | - | The ember-cli UI object that can be used to write to stdout.
+`appConfig` | `config/environment.js` | The application configuration.
+`deployConfig` | `config/deploy.js` | The deploy configuration file.
+`pluginConfig` | stored in `config/deploy.js` | The plugin's configuration. See configuring section below.
+
+
+Note: We can provide a default ``init`` that sets these properties.
+Authors can override if needed.
+
+### Plugins are singletons
+
+Every plugin is a singleton that is created when the deployment pipeline
+starts and torn down after the last pipeline step. The reasoning
+here is it allows the plugin to maintain state between different hooks.
+For example:
+
+```javascript
+TimerPlugin = DeployPlugin.extend({
+  startAt: null,
+
+  hooks: {
+    willDeploy: function() {
+      this.startAt = new Date();
+    },
+
+    didDeploy: function() {
+      console.log('deploy took %s', new Date() - this.startAt);
+    }
+  }
+});
+```
+
+### Returning from hooks
+
+Hook functions can return a promise to block the deployment pipeline.
+Since most deployment involves some sort of IO it makes senses that most
+plugins will want an async function to complete before continuing to the
+next step. If a plugin does not return a promise the next pipeline step
+begins running.
+
+Additionally, plugins can return a POJO, directly or from a promise,
+with data about what happened. Future plugins can access that data,
+more on that later.
+
```

Comment:
Does it make sense to make a plugin HAVE to return a Promise?  Seems like it might simplify things for us and for the people writing plugins if there is only one thing they can return.  We could use something similar to `ember-cli` Tasks where a plugin can either:

- resolve with undefined or a POJO
- reject with an Error instance if the error is unhandled or;
- reject with undefined if it was handled.